### PR TITLE
Additional features for CSW QGIS plugin

### PR DESCRIPTION
**Pull request by vesnikos**
- Added ability to get bbox extension from any loaded layer in qgis.
- Added ability to get bbox using geolocation services.
- UI tweaks to accommodate the new features. (UI was updated with Designer)

### CODE REVIEW

**Comment by tomkralidis** [comment]
cc @rduivenvoorde @alexbruy @kalxas 

@vesnikos thanks for the PR, these are amazing features!  Comments/suggestions:
- filter by layer extent: add an iterm to the "Layer" combobox called "Filter by Layer Extent", which, when selected, does nothing.  If I have QGIS open with one layer, I am currently given no choice but to have it selected
- geocoding: I'm confused about the workflow: do you select a placename, hit enter, then search again?  When clicking "Search" does this include the placename entered?  We should ensure this is clear/fluid (@rduivenvoorde may have suggestions here)
- can we take a stab at documenting these features and the added setting in https://github.com/geopython/MetaSearch/blob/master/docs/howto.rst?
- can we ensure the travis-ci tests pass (see https://travis-ci.org/geopython/MetaSearch/builds/36027089)?  The test defs at https://github.com/geopython/MetaSearch/blob/master/.travis.yml do some link checking as well as pep8 checking against all `.py` files

File: plugin/MetaSearch/dialogs/maindialog.py [hunk 1]
```diff
@@ -392,6 +491,45 @@ def set_bbox_global(self):
...
+            # http://en.wikipedia.org/wiki/List_of_Google_domains
+            geolocator = GoogleV3(timeout=4, domain="maps.google.com")
+            geotype = "googlev3"
+
+        elif self.rbGeolocationService_OSM.isChecked():
+            geolocator = Nominatim(view_box=(-180, -90, 180, 90), timeout=4)
+            geotype = "nominatim"
+
+        try:
+            location = geolocator.geocode(self.leWhere.text())
```
Comment (tomkralidis) [inline plugin/MetaSearch/dialogs/maindialog.py:558]: How are multiple results handled? Is the first result used? Perhaps we can provide a list of results to the user so they can choose, before setting the bbox?

File: plugin/MetaSearch/dialogs/maindialog.py [hunk 2]
```diff
@@ -392,6 +491,45 @@ def set_bbox_global(self):
...
+            # hackish way parsing results
+            maxx, maxy, minx, miny = ullr
+
+            # set radius of BBox
+            self.leNorth.setText(str(maxy))
+            self.leSouth.setText(str(miny))
+            self.leWest.setText(str(minx))
+            self.leEast.setText(str(maxx))
+        except (GeocoderTimedOut, GeocoderUnavailable):
+            self.leWhere.setText(u"Err: GeoQuerry Quota Exceeded")
```
Comment (tomkralidis) [inline plugin/MetaSearch/dialogs/maindialog.py]: We need to have strings like this translatable with `self.tr()` so it can be picked up by Transifex.

File: plugin/MetaSearch/dialogs/maindialog.py (hunk 1 above)
Comment (vesnikos) [inline plugin/MetaSearch/dialogs/maindialog.py:558]: The geolocator always returns one result; the most relevant one. After the user hits return, the where lineedit widget updates itself with the name and populates the bounding box. If the user is happy with the provided location then he can go ahead with the search function. If not he can always refine the location string with more information, e.g. by adding a country. 

It is possible for the geolocator to return n locations, so a user might choose from it. I think i can implement if it helps 

It is possible to pass an extra argument to bias the geolocator on where to focus if one wants - the bias also takes care the languange of the returned location. 

eg if you set the domain for google geolocator to "maps.google.gr" the returned string would be in greek. Its easy to implement such functionality, but i haven't thought a workable approach to that yet.

Comment (tomkralidis) [inline plugin/MetaSearch/dialogs/maindialog.py:558]: For example, what happens if one searches for 'Waterloo', there are many examples from many countries.  Maybe presenting a list is best, limited to top 10 results?

We should ensure global search to keep things simple for now.

**Comment by kalxas** [comment]
@vesnikos will you issue a new PR?
I think the review process is going well.

**Comment by vesnikos** [comment]
@kalxas Yes, I closed the pull request because of some stupid bugs. After I fix them I'll reopen the PL

**Comment by vesnikos** [comment]
@tomkralidis @kalxas 

Additionally from "Select from layer" method implemented a QCompleter to handle multiple locations returned from the geolocation service  

screenshot from 2014-09-26 18 09 50

File: plugin/MetaSearch/ui/maindialog - Copy.ui [hunk 3]
```diff
@@ -0,0 +1,557 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (tomkralidis) [inline plugin/MetaSearch/ui/maindialog - Copy.ui]: Is this file needed? Can we remove?

File: plugin/MetaSearch/ui/maindialog.ui [hunk 4]
```diff
@@ -524,16 +736,67 @@
...
   </layout>
  </widget>
+ <customwidgets>
+  <customwidget>
+   <class>mLineEdit</class>
+   <extends>QLineEdit</extends>
+   <header>customW.h</header>
+  </customwidget>
+ </customwidgets>
+ <tabstops>
```
Comment (tomkralidis) [inline plugin/MetaSearch/ui/maindialog.ui:762]: what do "tabstop" elements do?

File: plugin/MetaSearch/ui/customW/mLineEdit.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+###############################################################################
+__author__ = 'vesnikos'
```
Comment (tomkralidis) [inline plugin/MetaSearch/ui/customW/mLineEdit.py]: For all applicable files, Copyright/date is better place in the header (see above for format)

File: plugin/MetaSearch/ui/customW/__init__.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+###############################################################################
+__author__ = 'vesnikos'
+__date__ = '24/09/2014'
+
+
+from .mLineEdit import *
```
Comment (tomkralidis) [inline plugin/MetaSearch/ui/customW/__init__.py]: Is this required? Can we stay away from this type of import? (relative imports, and import *)

File: plugin/MetaSearch/ui/maindialog.ui (hunk 4 above)
Comment (vesnikos) [inline plugin/MetaSearch/ui/maindialog.ui:762]: Not really sure; I think it controls the tab order.  QDesigner implemented them when I was arranging them so they would behave in a predictable way.

File: plugin/MetaSearch/ui/maindialog - Copy.ui (hunk 3 above)
Comment (vesnikos) [inline plugin/MetaSearch/ui/maindialog - Copy.ui]: Yes; It shouldn't be there - still learning how to control git

**Comment by tomkralidis** [comment]
FYI note MetaSearch has now been merged into core QGIS, and any pull requests/changes can go through QGIS.  Thanks.
